### PR TITLE
[BEAM-759] Make TestPipeline.run fail when the underlying execution fails.

### PR DESCRIPTION
**Pull request by aaltay**
Also, DataflowRunner will log the last error from its wait_until_finish
method.

Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by aaltay** [comment]
R: @sb2nov

**Comment by sb2nov** [comment]
LGTM

**Comment by aaltay** [comment]
Thank you.

R: @robertwb

**Comment by robertwb** [comment]
I think it'd be a nicer user experience to raise the original exception, and possibly not just for the TestPipeline. I think this came up before, but in a language like Python (or Java) exceptions are the way to indicate abnormal behavior, and failing to check the status is error prone. Perhaps you could file a JIRA about this and add a TODO? 

(BTW, did we not have any tests that, say, errors in DoFns were correctly being propagated?)

**Comment by aaltay** [comment]
@robertwb Thank you. I agree with you. Returned the exception to its previous state and added a TODO there with a JIRA issue. 

We were checking for error propagation only with the DirectRunner and that was still raising the exception. This changes unifies both runners again. (Still keeping the assert in the TestPipeline in case a future runner only sets the state to failed without raising an exception.)

PTAL.

**Comment by robertwb** [comment]
Looks good, thanks.

File: sdks/python/apache_beam/runners/dataflow_runner.py [hunk 1]
```diff
@@ -718,3 +720,11 @@ def __str__(self):
 
   def __repr__(self):
     return '<%s %s at %s>' % (self.__class__.__name__, self._job, hex(id(self)))
+
+
+class DataflowRuntimeException(Exception):
+  """Indicates an error has occurred in running this pipeline."""
+
+  def __init__(self, msg, result):
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/dataflow_runner.py]: This needs a state arg according to the call above. (Thanks lint.)

Comment (aaltay) [inline sdks/python/apache_beam/runners/dataflow_runner.py]: Fixed this. Also added a unit test to check for the exception.

**Comment by robertwb** [comment]
LGTM

**Comment by aaltay** [comment]
Thank you.

**Pull request closed without merging**